build/net/src/lib.rs

```diff
@@ -82,23 +98,129 @@ pub fn load_net_config() -> Result<NetConfig> {
     Ok(cfg)
 }
 
-pub fn generate_vlan_consts(
+pub fn generate_port_consts(
     config: &NetConfig,
     mut out: impl std::io::Write,
 ) -> Result<(), std::io::Error> {
-    let vlan = config.vlan.unwrap();
-    let end = vlan.start + vlan.count;
-    if end > 0xFFF {
-        panic!("Invalid VLAN range (must be < 4096)");
+    let ports = config
+        .vlans
+        .values()
+        .map(|v| v.port)
+        .collect::<BTreeSet<_>>()
+        .len();
+    assert!(ports <= 2);
+    let port_count = ports.max(1);
+    let s = quote! {
+        pub const PORT_COUNT: usize = #port_count;
+    };
+    writeln!(out, "{s}")
+}
+
+fn check_vlan_config(config: &NetConfig) {
+    if config.vlans.values().any(|v| v.vid > 0xFFF) {
+        panic!("Invalid VLAN VID (must be < 4096)");
     }
-    writeln!(
-        out,
-        "
-pub const VLAN_RANGE: core::ops::Range<u16> = {:#x}..{:#x};
-pub const VLAN_COUNT: usize = {};
-",
-        vlan.start, end, vlan.count
-    )
+    for (k1, v1) in &config.vlans {
+        for (k2, v2) in &config.vlans {
+            if k1 != k2 && v1.vid == v2.vid {
+                panic!("Duplicate VID {} (in {k1} and {k2})", v1.vid);
+            }
+        }
+    }
+}
+
+pub fn generate_vlan_consts(
+    config: &NetConfig,
+    mut out: impl std::io::Write,
+) -> Result<(), std::io::Error> {
+    check_vlan_config(config);
+
+    let vlan_count = config.vlans.len();
+    let vids = config.vlans.values().map(|cfg| cfg.vid).collect::<Vec<_>>();
+    let s = quote! {
+        #[allow(unused)]
+        pub const VLAN_COUNT: usize = #vlan_count;
```

Comment:
out of curiosity: what's the motivation for having a separate constant for this, instead of just using `VLAN_VIDS.len()`?

Reply:
Historical inertia; I switched to using `VLanId::LENGTH` (which is generated from `enum_map::Enum`)